### ISSUE TITLE
Scalacp emits wrong symbols for Java references

### ISSUE THREAD

**Issue by olafurpg** [scalameta/scalameta#1491]
For the signature `  val f: java.util.Map.Entry[Int, Int]` metacp produces a wrong java symbol
```diff
- metacp
+ metac
_empty_.a.f(). => val method f: : Entry[Int, Int]
-  Entry => java.util.Map.Entry#
+  Entry => java.util.Map#Entry#
```

**Comment by olafurpg** [comment]
Hit on this again
```diff
object a {
  def lookup(handle: MethodHandles.Lookup): MethodHandles.Lookup = handle
}
- java.lang.invoke.MethodHandles.Lookup#
+ java.lang.invoke.MethodHandles#Lookup#
```

**Comment by olafurpg** [comment]
I dug a bit into this and it's not clear how to best fix this. The symbol for java references is an ExternalSymbol that gets wrongly assigned the kind `OBJECT`

```scala
scala.meta.internal.scalacp.Scalacp.SymbolOps#descriptor:435 sym: ExternalSymbol(
  "Entry",
  Some(
    ExternalSymbol(
      "Map",
      Some(
        ExternalSymbol(
          "util",
          Some(ExternalSymbol("java", None, Entry(38, 10, 0 bytes))),
          Entry(41, 10, 0 bytes)
        )
      ),
      Entry(44, 10, 0 bytes)
    )
  ),
  Entry(42, 9, 0 bytes)
)
scala.meta.internal.scalacp.Scalacp.SymbolOps#descriptor:436 skind(sym): CLASS
scala.meta.internal.scalacp.Scalacp.skind:115 sym.entry: Entry(44, 10, 0 bytes)
scala.meta.internal.scalacp.Scalacp.skind:116 sym.entry.entryType: 10
scala.meta.internal.scalacp.Scalacp.SymbolOps#descriptor:437 skind(sym.parent.get): OBJECT
```

The relevant piece of code that needs fixing is https://github.com/scalameta/scalameta/blob/0b41ec6a0cac00eef35fc043a74e6cffda12afdc/semanticdb/metacp/src/main/scala/scala/meta/internal/scalacp/Scalacp.scala#L111
The problem is that ExternalSymbol does not implement `.isJava`, the implementation of `hasFlag` returns false in all cases. Any idea how to fix this @xeno-by ? This is the most frequent issue I've encountered while implementing the type printer, causing problems in 16 classpath entries out of 87.

**Comment by xeno-by** [comment]
@olafurpg No idea. I've been cargo culting around external symbols, but it seems that we need to get to the bottom of it and figure out how they really work.

**Comment by xeno-by** [comment]
**Example**

Let's say we have a class Test defined as follows:

```scala
class Test {
  def foo: C.D = ???
}
```

Also, let's say we compile this class twice - once with a java definition for D:

```java
class C {
  public static class D {
  }
}
```

And once with a scala definition for D:

```scala
object C {
  class D
}
```

**Analysis**

According to the SemanticDB specification, in the first case D must be modelled as `_empty_.C#D`, whereas in the second case D must be modelled as `_empty_.C.D`, which means that SemanticDB generators must be able to distinguish the two definitions of D.

The problem is that it is impossible to distinguish those definitions by looking into the classfiles alone. In fact, not just ScalaSignatures for those two definitions are the same, but the classfiles themselves are identical.

Metac can emit correct symbols, because it has a symbol table, i.e. something that take a symbol and return symbol metadata. Looking into corresponding metadata reveals that the first D comes from java and the second D comes from scala, which is enough to produce correct symbols.

However, metacp doesn't have a symbol table. It is basically a `.class` file to `.semanticdb` file. This leads to the problem describeed in this ticket.

**Comment by xeno-by** [comment]
While it is certainly possible to add a symbol table to metacp, I think there are benefits to keeping it dumb. For one, dumber software is easier to maintain. Another reason is performance and amenability to distributed processing (if metacp needs a full classpath, then it's much harder to distribute).

Therefore, I propose that we consider changing the Java symbol format. Non-static inner classes would still retain their `C#D` format, whereas static inner classes would change to `C.D`. This will make metacp behave correctly out of the box, will require small but manageable changes in metac, and best of all would still remain syntactic-only, so that mtags can continue emitting SemanticDBs.

**Comment by olafurpg** [comment]
> Non-static inner classes would still retain their C#D format, whereas static inner classes would change to C.D. This will make metacp behave correctly out of the box, will require small but manageable changes in metac, and best of all would still remain syntactic-only

I'm afraid this will not work for Java signatures that reference external static inner classes. For example,

```java
class A {
  public java.util.Map.Entry<String, String> entry = null
}
```

If metacp is stateless, we have no way to determine whether `Map.Entry` is a static class or not. I don't see a way around making metacp stateful if we customize the symbol format based on semantics.

I looked into what IntelliJ Scala does and turns out that their fork of scalap has `ExternalSymbol.isObject: Boolean`. That boolean bit is computed from the tag `EXTMODCLASSref`

https://github.com/JetBrains/intellij-scala/blob/67c0fa05b7e9a4b730b80062131cae6085c2c2bf/scala/decompiler/src/org/jetbrains/plugins/scala/decompiler/scalasig/Parser.scala#L171-L172

I can't see references to `EXTMODCLASSref` in scalap but scala-reflect and scala-compiler reference it.

I propose we fork the IntelliJ Scala signature parser and see if that solves this problem. If successful, we profit from

- solving this problem, which is currently blocking Scalafix v0.6 (https://github.com/scalacenter/scalafix/issues/761)
- keeping metacp state-less
- better metacp performance (no more parser combinators)
- years of battle-testing in IntelliJ Scala

**Comment by olafurpg** [comment]
@xeno-by observed that classfiles embed metadata about which referenced symbols are static. For example,

```scala
class C {
  def m: java.util.Map.Entry[String, String] = ???
}
```
Running `java -v C` with verbose mode

```
InnerClasses:
     public static #13= #10 of #12; //Entry=class java/util/Map$Entry of class java/util/Map
```

From the JVM spec:

> 4.7.6. The InnerClasses Attribute
The InnerClasses attribute is a variable-length attribute in the attributes table of a ClassFile structure (§4.1).
> 
> If the constant pool of a class or interface C contains at least one CONSTANT_Class_info entry (§4.4.1) which represents a class or interface that is not a member of a package, then there must be exactly one InnerClasses attribute in the attributes table of the ClassFile structure for C.

Thanks to this observation, the problem I raised in my previous comment is solved and there is no need to migrate to the IntelliJ Scala decompiler. I've opened #1629 fixing this ticket.

I still believe migrating to the IntelliJ Scala decompiler is a good idea for the improved performance benefits. However, this we can do after v4.0 is released since the migration will not affect the spec.

**Comment by olafurpg** [comment]
Fixed in #1638.
